From 50dd72329fc1268ca6e0808f906aee7d2dd7efcc Mon Sep 17 00:00:00 2001
From: Alejandro del Castillo <alejandro.delcastillo@ni.com>
Date: Fri, 31 Aug 2018 17:25:44 +0300
Subject: [PATCH] libsolv_solver_execute_transaction: abort transaction on
 first failure

If package A depends on B, and B is corrupted, if you install A, opkg
will try to install B, fail and continue installing A. Change logic to
abort the transaction on first failure. This is consistent with the
internal solver behavior.

Signed-off-by: Alejandro del Castillo <alejandro.delcastillo@ni.com>
---
 libopkg/solvers/libsolv/opkg_solver_libsolv.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/libopkg/solvers/libsolv/opkg_solver_libsolv.c b/libopkg/solvers/libsolv/opkg_solver_libsolv.c
index d261b40..0a4832d 100644
--- a/libopkg/solvers/libsolv/opkg_solver_libsolv.c
+++ b/libopkg/solvers/libsolv/opkg_solver_libsolv.c
@@ -819,6 +819,10 @@ static int libsolv_solver_execute_transaction(libsolv_solver_t *libsolv_solver)
             switch (typeId) {
             case SOLVER_TRANSACTION_ERASE:
                 ret = opkg_remove_pkg(pkg);
+                if (ret) {
+                    err = -1;
+                    goto CLEANUP;
+                }
                 break;
             case SOLVER_TRANSACTION_DOWNGRADE:
             case SOLVER_TRANSACTION_REINSTALL:
@@ -841,6 +845,10 @@ static int libsolv_solver_execute_transaction(libsolv_solver_t *libsolv_solver)
                                  pkg->name, pkg->version, pkg->dest->name);
                 }
                 ret = opkg_install_pkg(pkg, 0);
+                if (ret) {
+                    err = -1;
+                    goto CLEANUP;
+                }
                 break;
             case SOLVER_TRANSACTION_UPGRADE:
                 old = pkg_hash_fetch_installed_by_name(pkg->name);
@@ -871,14 +879,15 @@ static int libsolv_solver_execute_transaction(libsolv_solver_t *libsolv_solver)
                 }
 
                 ret = opkg_install_pkg(pkg, 1);
+                if (ret) {
+                    err = -1;
+                    goto CLEANUP;
+                }
                 break;
             case SOLVER_TRANSACTION_IGNORE:
             default:
-                ret = 0;
                 break;
             }
-            if (ret)
-                err = -1;
         }
     }
 
-- 
2.18.0

